File: teams-ai-kit/python/teams_ai_kit/state.py
"""TurnState — conversation/user/temp scopes backed by pluggable storage."""

from __future__ import annotations

import copy
from typing import Any, Dict

from botbuilder.core import TurnContext


class TurnState:
    def __init__(self) -> None:
        self.conversation: Dict[str, Any] = {}
        self.user: Dict[str, Any] = {}
        self.temp: Dict[str, Any] = {}


class MemoryStorage:
    """In-memory storage. Swap in any Storage implementation for persistence."""

    def __init__(self) -> None:
        self._memory: Dict[str, Dict[str, Any]] = {}

    async def read(self, keys):
        return {key: copy.deepcopy(self._memory[key]) for key in keys if key in self._memory}

    async def write(self, changes: Dict[str, Dict[str, Any]]) -> None:
        for key, value in changes.items():
            self._memory[key] = copy.deepcopy(value)

    async def delete(self, keys) -> None:
        for key in keys:
            self._memory.pop(key, None)


def _storage_keys(ctx: TurnContext):
    channel = ctx.activity.channel_id or "unknown"
    conversation = (ctx.activity.conversation and ctx.activity.conversation.id) or f"{channel}/none"
    user = (ctx.activity.from_property and ctx.activity.from_property.id) if hasattr(ctx.activity, "from_property") else None
    if user is None:
        user = getattr(ctx.activity, "from_id", None) or "anonymous"
    return f"{channel}/{conversation}", f"{channel}/{user}"


async def load_state(ctx: TurnContext, storage: MemoryStorage) -> TurnState:
    conversation_key, user_key = _storage_keys(ctx)
    stored = await storage.read([conversation_key, user_key])
    state = TurnState()
    state.conversation = stored.get(conversation_key, {})
    state.user = stored.get(user_key, {})
    return state


async def save_state(ctx: TurnContext, storage: MemoryStorage, state: TurnState) -> None:
    conversation_key, user_key = _storage_keys(ctx)
    changes: Dict[str, Dict[str, Any]] = {}
    if state.conversation:
        changes[conversation_key] = state.conversation
    if state.user:
        changes[user_key] = state.user
    await storage.write(changes)

File: teams-ai-kit/node/src/state.ts
/**
 * TurnState — scoped state for each turn, like the classic TurnState model:
 * `conversation` (shared per conversation), `user` (per user), `temp` (this turn).
 */

import type { TurnContext } from "botbuilder";

export type Scope = Record<string, any>;

export type TurnState = {
  conversation: Scope & { history?: { role: "user" | "assistant"; text: string }[] };
  user: Scope;
  temp: Scope & { locale?: string; input?: string; authTokens?: Scope };
};

export interface Storage {
  read(keys: string[]): Promise<Record<string, Scope>>;
  write(changes: Record<string, Scope>): Promise<void>;
  delete(keys: string[]): Promise<void>;
}

/** In-memory storage (default). Swap in any Storage implementation for persistence. */
export class MemoryStorage implements Storage {
  private memory = new Map<string, Scope>();

  async read(keys: string[]): Promise<Record<string, Scope>> {
    const result: Record<string, Scope> = {};
    for (const key of keys) {
      const value = this.memory.get(key);
      if (value) result[key] = JSON.parse(JSON.stringify(value));
    }
    return result;
  }

  async write(changes: Record<string, Scope>): Promise<void> {
    for (const [key, value] of Object.entries(changes)) {
      this.memory.set(key, JSON.parse(JSON.stringify(value)));
    }
  }

  async delete(keys: string[]): Promise<void> {
    for (const key of keys) this.memory.delete(key);
  }
}

function storageKeys(ctx: TurnContext): { conversation: string; user: string } {
  const channel = ctx.activity.channelId || "unknown";
  const conversation = ctx.activity.conversation?.id || `${channel}/no-conversation`;
  const user = ctx.activity.from?.id || "anonymous";
  return {
    conversation: `${channel}/${conversation}`,
    user: `${channel}/${user}`,
  };
}

/** Load the three scopes for this turn from storage. */
export async function loadState(ctx: TurnContext, storage: Storage): Promise<TurnState> {
  const keys = storageKeys(ctx);
  const stored = await storage.read([keys.conversation, keys.user]);
  return {
    conversation: stored[keys.conversation] ?? {},
    user: stored[keys.user] ?? {},
    temp: {},
  };
}

/** Persist the durable scopes (conversation + user; temp is per-turn only). */
export async function saveState(ctx: TurnContext, storage: Storage, state: TurnState): Promise<void> {
  const keys = storageKeys(ctx);
  const changes: Record<string, Scope> = {};
  if (Object.keys(state.conversation).length > 0) changes[keys.conversation] = state.conversation;
  if (Object.keys(state.user).length > 0) changes[keys.user] = state.user;
  await storage.write(changes);
}
